Worked the sleep session and breastfeeding pages so that data can be …

NicoBir opened the pull request:
…entered and shown by the app.

NicoBir committed:
Worked the sleep session and breastfeeding pages so that data can be entered and shown by the app.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+    val hourEdit = findViewById<EditText>(R.id.breast_hour)
+    val lengthEdit = findViewById<EditText>(R.id.breast_length)
+    val babyEdit = findViewById<EditText>(R.id.breast_baby)
+    val motherEdit = findViewById<EditText>(R.id.breast_mother)
```
Did you check if this works? Views aren't available until `onCreate` is called. I recommend to checkout view bindings instead: https://developer.android.com/topic/libraries/view-binding

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
 import android.os.Bundle
+import android.text.TextUtils
+import android.view.View
+import android.widget.*
```
Don't use star imports, you can configure your IDE to avoid that.

jbiral replied:
This is still not resolved.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
         setContentView(R.layout.breast_layout)
+
+        val mBreastSpinner: Spinner =
+            findViewById(com.example.babytracking.R.id.spinner_breast);
```
Import `R` instead of using its fully qualified name

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+    val lengthEdit = findViewById<EditText>(R.id.breast_length)
+    val babyEdit = findViewById<EditText>(R.id.breast_baby)
+    val motherEdit = findViewById<EditText>(R.id.breast_mother)
+    var mBreast: Boolean = false;
```
Use an [enum](https://kotlinlang.org/docs/reference/enum-classes.html) to differentiate left and right. It's much clearer that way

jbiral replied:
don't use `m` prefixes, they're evil :)

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
 import androidx.appcompat.app.AppCompatActivity
 
-class BreastActivity: AppCompatActivity() {
+class BreastActivity : AppCompatActivity() {
```
Can you add a Kotlin doc block to explain what this screen is doing?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+                "Hour = " + hourText + ", \n Length = " + lengthText
+                        + ", \n Breast = " + breastText + ", \n Session for baby = " + babyText
+                        + ", \n Session for mother = " + motherText,
+                Toast.LENGTH_SHORT
```
This is too long to be in a lambda. Could you extract it to a function instead?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+                    override fun onNothingSelected(parent: AdapterView<*>?) {
+                        mBreast = false
+                    }
+                }
```
Could you extract this to an inner class instead?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+
+                    // Because AdapterView is an abstract class, onNothingSelected must be defined
+                    override fun onNothingSelected(parent: AdapterView<*>?) {
+                        mBreast = false
```
Is it possible to prevent having nothing selected? If not, maybe we should have an `UNKNOWN` value in the enum and not allow to click the button if not set

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+            mBreastSpinner.onItemSelectedListener =
+                object : AdapterView.OnItemSelectedListener {
+                    override fun onItemSelected(
+                        parent: AdapterView<*>, view: View, pos: Int, id: Long
```
Don't use `<*>`: specify the type instead (`String`)

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+                                mBreast = false // Left
+                            }
+
+                        }
```
Is this needed? Or can you just check [getSelectedItemId()](https://developer.android.com/reference/android/widget/AdapterView.html#getSelectedItemId()) when the button is clicked?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,85 @@
+                this@BreastActivity,
+                "Hour = " + hourText + ", \n Length = " + lengthText
+                        + ", \n Breast = " + breastText + ", \n Session for baby = " + babyText
+                        + ", \n Session for mother = " + motherText,
```
In Kotlin, you can do this much more nicely:

```kotlin
val userInput = 
  """
    Hour = $hourText
    Length = $lengthText
    Breast side = $breastText
    etc.
  """
```

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
 import android.os.Bundle
+import android.text.TextUtils
+import android.view.View
+import android.widget.*
```
Same here

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
-class SleepActivity: AppCompatActivity() {
+class SleepActivity : AppCompatActivity() {
+
+    var mLullaby: Boolean = false;
```
Please don't use `m` prefixes

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
+        val lengthEdit = findViewById<EditText>(R.id.sleep_length)
+        val qualityEdit = findViewById<EditText>(R.id.sleep_quality)
+        val mLullabySpinner: Spinner =
+            findViewById(com.example.babytracking.R.id.spinner_sleep_lullaby);
```
import `R`

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
+        val hourEdit = findViewById<EditText>(R.id.sleep_hour)
+        val lengthEdit = findViewById<EditText>(R.id.sleep_length)
+        val qualityEdit = findViewById<EditText>(R.id.sleep_quality)
+        val mLullabySpinner: Spinner =
```
No `m` annotation please :)

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.sleep_layout)
+        val saveButton = findViewById<Button>(R.id.sleep_button_save)
```
I think this form is more consistent with what you do elsewhere:
```kotlin
val saveButton: Button = findViewById(R.id.sleep_button_save)
```

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,83 @@
+                    // Because AdapterView is an abstract class, onNothingSelected must be defined
+                    override fun onNothingSelected(parent: AdapterView<*>?) {
+                        mLullaby = false
+                    }
```
If this is a yes/no input, what about using a checkbox?

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:hint="Enter Hour"
+            android:inputType="textCapWords" />
```
Should it be a time instead? You could use a [TimePicker](https://developer.android.com/reference/android/widget/TimePicker)

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+            android:text="At wich hour?" />
+
+        <EditText
+            android:id="@+id/breast_hour"
```
`time` is a bit more correct in English than hour

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:hint="Enter Time"
+            android:inputType="textCapWords" />
```
Should this be a number? In minutes? Is there a max value allowed? a minimum value? Can I input -1? 0? 3600 minutes?

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
 <?xml version="1.0" encoding="utf-8"?>
```
All strings and dimensions should be defined in xml files

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+            android:text="How was the session for the baby?" />
+
+        <EditText
+            android:id="@+id/breast_baby"
```
This id is not descriptive enough. Think about making the id clear so that if I only read the id, I know what the view is about. The same should apply for other ids as well.

jbiral replied:
Tip: you can use Shift + F6 to rename the id everywhere at once in Android Studio

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:hint="Enter Quality"
+            android:inputType="textCapWords" />
```
I don't know why you're using `textCapWords` everywhere, but it doesn't seem appropriate. It will capitalize each work and `Write Like This`. Check out the other [options](https://developer.android.com/reference/android/widget/TextView#attr_android:inputType) and pick the most appropriate given the type of input

jbiral commented on app/src/main/res/layout/breast_layout.xml:
```diff
@@ -1,14 +1,111 @@
+        android:layout_gravity="center_horizontal"
+        android:text="Save" />
 
 </LinearLayout>
```
nit: add a new line

jbiral commented on app/src/main/res/layout/sleep_layout.xml:
```diff
@@ -1,14 +1,90 @@
+        <TextView
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="At wich hour?" />
```
At what time?

jbiral commented on app/src/main/res/values/arrays.xml:
```diff
@@ -0,0 +1,22 @@
+     limitations under the License.
+-->
+<resources>
+    <!-- These are the options displayed in the gender drop-down Spinner -->
```
Why gender?

jbiral commented on app/src/main/res/values/arrays.xml:
```diff
@@ -0,0 +1,22 @@
+        <item>@string/left_breast_string</item>
+        <item>@string/right_breast_string</item>
+    </string-array>
+</resources>
```
Don't forget the empty line at the end. You might need to configure the IDE to do it for XMLs too

jbiral commented on app/src/main/java/com/example/babytracking/BabyActivity.kt:
```diff
@@ -1,11 +1,13 @@
 package com.example.babytracking
```
Don't forget to update the package in another PR (After this one, for example)

jbiral commented:
You should be pushing your fixes in here, not in a new PR.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+                    ampm = "PM"
+                }
+                else -> ampm = "AM"
+            }
```
This could be more readable if extracted into a helper function, what do you think?

jbiral replied:
We really shouldn't be storing this as a string. It should be converted to a String just for display, otherwise please store it as a `Long` timestamp.

Here, you can convert it to a timestamp using [`LocalDateTime`](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html) and [`ZonedDateTime`](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html):

```kt
val localDateTime = LocalDateTime.of(year, month, dayOfMonth, hour, minute) // You'll notice that you need to figure out how to have a user input the date too, as it's possible we'll be inputting an event for a previous day.
val zonedDateTime = ZonedDateTime.of(localDateTime, ZoneId.getSystemDefault())
val timestamp = zonedDateTime.toInstant().toEpochMilli()
```

There are many reasons why this is preferable. One of them being that String require more bits to be stored, so a timestamp will be more efficient. Additionally, the timestamp contains extra data like the date and timezone, and can be converted across timezones if needed. Last, but not least, it's much easier to sort events using a timestamp.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+
+        breastTimeClick()
+        breastSelection()
+        breastSave()
```
These function names aren't descriptive enough. Do you mean `setupTimePicker`? `setupBreastSpinner`? `setupSaveButton`?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+    }
+
+    private fun breastTimeClick() {
+        val timePicker = findViewById<TimePicker>(R.id.sleep_timepick_hour)
```
Why don't you import it directly like the other views?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+import kotlinx.android.synthetic.main.breast_layout.breast_button_save
+import kotlinx.android.synthetic.main.breast_layout.breast_length
+import kotlinx.android.synthetic.main.breast_layout.breast_quality_mother
+import kotlinx.android.synthetic.main.breast_layout.spinner_breast
```
While layout IDs typically use snake_case, the convention is to use camelCase for views that you will access in code by their IDs using Kotlin. This way, you keep your classes more Android-y.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+    }
+
+    private fun breastSelection() {
+        if (spinner_breast != null) {
```
Can this even be null? Why?

If it's truly an issue, consider returning early to avoid having too much nested code:

```kt
if (spinner_breast == null) {
  return
}
```

Or you could try using the Elvis `?:` operator:

```kt
spinner_breast ?: return
```

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+    }
+
+    private fun breastSave() {
+        breast_button_save.setOnClickListener {
```
I feel like it would be much better to extract the content of the `setOnClickListener` to another function, and then remove the `breastSave` function and directly set the click listener in `onCreate`:

```kt
  breast_button.setOnClickListener { saveInputs()  }
```

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+        breast_button_save.setOnClickListener {
+
+            val lengthText: String = breast_length.text.toString().trim()
+            val lenghtInt: Int = Integer.parseInt(lengthText)
```
nit: use the KT operators:

```kt
lengthText.toInt()
```

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+            Breast = $breast
+            Session for baby = $babyText
+            Session for mother = $motherText
+            """
```
the indenting seems off here

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+            val motherText = breast_quality_mother.text.toString().trim()
+            val hourVal = if (hour < 10) "0" + hour else hour
+            val minVal = if (minute < 10) "0" + minute else minute
+            val timeText = "$hourVal : $minVal $ampm"
```
See the part about using a timestamp rather than this. This logic is fine, but should be used only for display. Additionally, there are a ton of utility classes that can format a timestamp to a readable time / date for you, so you don't have to handle that logic yourself

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+                breast_length.error = "Time cannot exceed 60 minutes"
+                breast_length.requestFocus()
+                return@setOnClickListener
+            }
```
Would these actually ever happen given that the time picker doesn't allow that? I think it's fine to check for it, but having an error message might be overkill since it's unlikely to happen

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+            }
+
+            if (babyText.isEmpty()) {
+                breast_quality_baby.error = "Quality for baby required"
```
This isn't very clear?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,136 @@
+            }
+
+            if (motherText.isEmpty()) {
+                breast_quality_mother.error = "Quality for mother required"
```
Same here

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,95 @@
+            minute = pickerMinute
+
+            // AM_PM decider logic
+            when {
```
See comments in other file. Also, if you have the exact same logic in 2 places, could you consider sharing the logic somehow?

You could have a helper class to parse the hour / minute from the picker and return a timestamp, for example. So this class can be shared by both activities

jbiral commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,20 @@
+    <string name="sleep_qual_text">How was the baby\'s sleep?</string>
+    <string name="sleep_qual_edit">Enter quality of sleep</string>
+    <string name="sleep_lull_edit">Did the baby need a lullaby to sleep?</string>
 </resources>
```
Empty line?

jbiral commented:
I noticed you didn't address all the comments from last time. To make sure you've seen and addressed the comments, could you mark them as resolved when done? Thanks!

NicoBir committed:
Created a DatePicker to allow the input of the date.
Modified the TimePicker and the DatePicker so that they pop-up when needed without taking all of the screen.
Used a timestamp to save the data related to time.
Corrected some things.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+import java.sql.Timestamp
+import java.util.*
+
+/*
+This screen will allow to enter the data about the breastfeeding sessions of the baby.
+The first data required is the date and time when the session occurred.
+Those two are taken care of by the objects PickerSetup, which allow to pick the date and time,
+and TimeFormatter, which format the time so that we can read it.
+The other data that can be entered in this screen are the length of the session,
+at which breast did the baby drink, as well as how the session was for the baby and for the mother.
+*/
```
This isn't the right format for Kotlin doc, please read the official documentation: https://kotlinlang.org/docs/reference/kotlin-doc.html

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+import java.util.*
+
+/*
+This screen will allow to enter the data about the breastfeeding sessions of the baby.
```
The summary block (the first line) should be succinct: `Allows to keep track of breastfeeding sessions.`

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+
+/*
+This screen will allow to enter the data about the breastfeeding sessions of the baby.
+The first data required is the date and time when the session occurred.
+Those two are taken care of by the objects PickerSetup, which allow to pick the date and time,
+and TimeFormatter, which format the time so that we can read it.
+The other data that can be entered in this screen are the length of the session,
+at which breast did the baby drink, as well as how the session was for the baby and for the mother.
```
Use an unordered list to describe the different fields:

```kt
/**
 * Summary block
 *
 * <p> The following fields are required:
 * <ul>
 *   <li> Date and time of the session.
 *.   <li> Duration of the session. (Etc.)
 * </ul>
 */

keep it short but descriptive, people don't read docs when they're too extensive.
```

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+class BreastActivity : AppCompatActivity() {
+
+    private var breast: String = ""
+    private var cal: Calendar = Calendar.getInstance()
```
why not a `val`?

jbiral replied:
please don't use abbreviations: `calendar` is better.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
         super.onCreate(savedInstanceState)
         setContentView(R.layout.breast_layout)
+
+        PickerSetup.setupDatePicker(this, breast_Date, cal)
```
`breast_Date` isn't camel case. Use the following case: `breastDate` (same for other IDs)

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+    }
+
+    private fun setupBreastSpinner() {
+
```
Looks like an unnecessary empty line? Maybe add an empty line after the return instead?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+        breast_Spinner.onItemSelectedListener =
+            object : AdapterView.OnItemSelectedListener {
+                override fun onItemSelected(
+                    parent: AdapterView<*>, view: View, pos: Int, id: Long
```
Why do you keep a star? `<*>`

Can you use the real type instead?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+    @Suppress("DEPRECATION")
+    @SuppressLint("NewApi")
+    private fun saveInput() {
+
```
remove this empty line

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+    }
+
+
+    @Suppress("DEPRECATION")
+    @SuppressLint("NewApi")
```
This is a bad practice: typically, you don't want to use deprecated APIs unless there are no other way around.

`NewApi` is even worse: it means you're using an API that's not supported by your `minSdk`, so it will crash on those devices. Please avoid doing so, and follow the tips at https://developer.android.com/training/basics/supporting-devices/platforms

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+    @SuppressLint("NewApi")
+    private fun saveInput() {
+
+        val lengthText: String = breast_Length.text.toString().trim()
```
this is not a great variable name: `lengthText` is meaningless and remove the context of which length we're referring to. Please use descriptive variable names, such as `sessionLength`. By the way, your ID could be better too: we know it's the activity for tracking breastfeeding sessions, and `breastLength` isn't descriptive at all because it seems like you're referring to the length of a breast rather than the session.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+        val lengthText: String = breast_Length.text.toString().trim()
+
+        if (lengthText.isEmpty()) {
+            breast_Length.error = "Length of breastfeeding required"
```
length is typically a distance. For time, please talk about duration.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+        val timestamp: Timestamp = TimeFormatter.makeTimeStamp(cal)
+        val date = Date(timestamp.time)
+
+        if (lengthInt <= 0) {
+            sleep_Length.error = "Time cannot be negative or null"
+            sleep_Length.requestFocus()
+            return
+        } else if (lengthInt > 24) {
+            sleep_Length.error = "Time cannot exceed 24 hours"
+            sleep_Length.requestFocus()
+            return
+        }
```
A session is never gonna take a day. The user should be able to input minutes rather than hours.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+        val sessionQualForBabyText: String = breast_Quality_Baby.text.toString().trim()
+        val sessionQualForMotherText: String = breast_Quality_Mother.text.toString().trim()
+
+        val timestamp: Timestamp = TimeFormatter.makeTimeStamp(cal)
```
This is very unclear: the calendar hides all the logic about which timestamp it is (and your variable name doesn't help). Could you store the timestamp in this class instead of a calendar?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+
+        val lengthText: String = breast_Length.text.toString().trim()
+
+        if (lengthText.isEmpty()) {
+            breast_Length.error = "Length of breastfeeding required"
+            breast_Length.requestFocus()
+            return
+        }
+
+        val lengthInt: Int = lengthText.toInt()
+        val sessionQualForBabyText: String = breast_Quality_Baby.text.toString().trim()
+        val sessionQualForMotherText: String = breast_Quality_Mother.text.toString().trim()
+
+        val timestamp: Timestamp = TimeFormatter.makeTimeStamp(cal)
+        val date = Date(timestamp.time)
+
+        if (lengthInt <= 0) {
+            sleep_Length.error = "Time cannot be negative or null"
+            sleep_Length.requestFocus()
+            return
+        } else if (lengthInt > 24) {
+            sleep_Length.error = "Time cannot exceed 24 hours"
+            sleep_Length.requestFocus()
+            return
+        }
+
+        if (sessionQualForBabyText.isEmpty()) {
+            sleep_Quality.error = "Quality of the session for the baby required"
+            sleep_Quality.requestFocus()
+            return
+        }
+
+        if (sessionQualForMotherText.isEmpty()) {
+            sleep_Quality.error = "Quality of the session for the mother required"
+            sleep_Quality.requestFocus()
+            return
+        }
```
All of this would be best in a `validate` function which returns false if the data isn't valid.

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+        val userInput =
+            """
+            TimeStamp = $timestamp
+            Date = ${TimeFormatter.formatDate(date)}
+            Hour = ${TimeFormatter.formatHour(timestamp.hours, timestamp.minutes)}
+            Length = $lengthInt
```
Which date? Which hour? (do you mean time?)

Duration of the session isntead of length

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
+            Hour = ${TimeFormatter.formatHour(timestamp.hours, timestamp.minutes)}
+            Length = $lengthInt
+            Breast = $breast
+            Session for baby = $sessionQualForBabyText
+            Session for mother = $sessionQualForMotherText
```
These qualifiers and variable names aren't meaningful enough. Do you mean quality of the session?

jbiral commented on app/src/main/java/com/example/babytracking/BreastActivity.kt:
```diff
@@ -1,11 +1,134 @@
\ No newline at end of file
+}
+
+
+
+
```
Delete

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+
+/*
+This object allows the creation of date and time pickers.
+*/
```
Most comments from the other file apply here too

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+
+     fun setupDatePicker(context: Context, date_TextView: TextView, cal: Calendar) {
+
+        date_TextView!!.text = "--/--/----"
```
never ever use `!!`, it means you're doing something wrong

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+*/
+object PickerSetup {
+
+     fun setupDatePicker(context: Context, date_TextView: TextView, cal: Calendar) {
```
use camelCase please

jbiral replied:
You should not set a text view from a helper class. Can you return the string to show instead?

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+                cal.set(Calendar.YEAR, yearDate)
+                cal.set(Calendar.MONTH, monthOfYear)
+                cal.set(Calendar.DAY_OF_MONTH, dayOfMonth)
+                date_TextView!!.text = TimeFormatter.formatDate(cal.time)
```
!!

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+                date_TextView!!.text = TimeFormatter.formatDate(cal.time)
+            }
+
+        date_TextView!!.setOnClickListener {
```
!!

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+
+        val dateSetListener =
+            DatePickerDialog.OnDateSetListener { _, yearDate, monthOfYear, dayOfMonth ->
+                cal.set(Calendar.YEAR, yearDate)
+                cal.set(Calendar.MONTH, monthOfYear)
+                cal.set(Calendar.DAY_OF_MONTH, dayOfMonth)
```
Do you really need to use a Calendar? Why not returning a timestamp? If you need to return multiple fields, feel free to create a data class that contains the return values that you need.

jbiral commented on app/src/main/java/com/example/babytracking/PickerSetup.kt:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+     fun setupTimePicker(context: Context, time_TextView: TextView, cal: Calendar) {
```
same comments here

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,89 @@
+how was their sleep and whether or not a lullaby was required to make them sleep.
+*/
+
+class SleepActivity : AppCompatActivity() {
```
Same comments apply here

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,89 @@
+
+class SleepActivity : AppCompatActivity() {
+
+    private var cal = Calendar.getInstance()
```
val, but I think we won't keep the calendar.

jbiral commented on app/src/main/java/com/example/babytracking/SleepActivity.kt:
```diff
@@ -1,11 +1,89 @@
+    }
+
+
+    @Suppress("DEPRECATION")
+    @SuppressLint("NewApi")
```
please don't do this

jbiral commented on app/src/main/java/com/example/babytracking/TimeFormatter.kt:
```diff
@@ -0,0 +1,62 @@
+import java.time.ZoneId
+import java.time.ZonedDateTime
+import java.util.*
+/*
```
empty line before this

jbiral commented on app/src/main/java/com/example/babytracking/TimeFormatter.kt:
```diff
@@ -0,0 +1,62 @@
+/*
+This object allows the formatting of the different time variables so that they can be accessed and read.
+*/
+object TimeFormatter {
```
same comments apply here

jbiral commented on app/src/main/java/com/example/babytracking/TimeFormatter.kt:
```diff
@@ -0,0 +1,62 @@
+
+
+    fun formatHour(hourInput: Int, minuteInput: Int): String {
+        var hour: Int = hourInput
+        val minute: Int = minuteInput
+        var ampm: String = "AM"
+        when {
+            hour == 0 -> {
+                hour += 12
+                ampm = "AM"
+            }
+            hour == 12 -> ampm = "PM"
+            hour > 12 -> {
+                hour -= 12
+                ampm = "PM"
+            }
+            else -> ampm = "AM"
+        }
+
+        val hourFormat: String = if (hour < 10) "0" + hour else "$hour"
+        val minuteFormat: String = if (minute < 10) "0" + minute else "$minute"
+        val formatText: String = "$hourFormat:$minuteFormat $ampm"
+
+        return formatText
```
TimeFormat can do this for you

jbiral commented on app/src/main/java/com/example/babytracking/TimeFormatter.kt:
```diff
@@ -0,0 +1,62 @@
+
+    @SuppressLint("NewApi")
+    fun makeTimeStamp(cal: Calendar): Timestamp {
+        val localDateTime = LocalDateTime.of(
+            cal.get(Calendar.YEAR),
+            cal.get(Calendar.MONTH) + 1,
+            cal.get(Calendar.DAY_OF_MONTH),
+            cal.get(Calendar.HOUR_OF_DAY),
+            cal.get(Calendar.MINUTE)
+        )
+        val zonedDateTime = ZonedDateTime.of(localDateTime, ZoneId.systemDefault())
+        val timestamp = Timestamp(zonedDateTime.toInstant().toEpochMilli())
+
+        return timestamp
```
Please don't use calendar, it's a buggy class and make your code more complex than it should be
